### ISSUE TITLE
torch.multinomial crashes with SIGSEGV on MPS for (larger) input tensors

### ISSUE THREAD

**Issue by zap150** [pytorch/pytorch#178579]
### Description

torch.multinomial crashes the Python process with SIGSEGV (exit code 139) when called on a large enough MPS tensor., The crash is deterministic and reproducible on a fixed machine. Below the threshold the function works correctly. Reproducible with both 2.10 and 2.11.

### Reproducer

```python
import os
import torch

print(f"PyTorch version: {torch.__version__}")
print(f"MPS available:   {torch.backends.mps.is_available()}")
print()

device = torch.device("mps")
gen = torch.Generator(device=device)
gen.manual_seed(42)

for exp in range(10, 25):  # 1024 .. 16M
    n = 2**exp
    w = torch.rand(n, device=device)
    print(f"n=2^{exp:2d} = {n:>9,}: testing...", end="  ", flush=True)
    idx = torch.multinomial(w, 100, replacement=True, generator=gen)
    torch.mps.synchronize()
    print(f"OK  (sample[0]={idx[0].item()})", flush=True)

print("\nNo crash found in tested range.")
os._exit(0)
```

### Output

```
PyTorch version: 2.10.0
MPS available:   True

n=2^10 =     1,024: testing...  OK  (sample[0]=846)
n=2^11 =     2,048: testing...  OK  (sample[0]=128)
n=2^12 =     4,096: testing...  OK  (sample[0]=2417)
n=2^13 =     8,192: testing...  OK  (sample[0]=5361)
n=2^14 =    16,384: testing...  OK  (sample[0]=16348)
n=2^15 =    32,768: testing...  OK  (sample[0]=881)
n=2^16 =    65,536: testing...  OK  (sample[0]=46760)
n=2^17 =   131,072: testing...  zsh: segmentation fault  python test_mps_multinomial_bug.py
```

### Versions

```
PyTorch version: 2.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.3.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.4.2)
CMake version: version 3.31.7
Libc version: N/A

Python version: 3.11.9 (v3.11.9:de54cf5be3, Apr  2 2024, 07:12:50) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-26.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] numpy==2.4.3
[pip3] torch==2.10.0
[pip3] torchaudio==2.10.0
[pip3] torchvision==0.25.0
[conda] Could not collect
```

cc @pbelevich @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

**Comment by Isalia20** [comment]
Will look into it in the upcoming week

**Comment by zap150** [comment]
@Isalia20 has this indeed been merged? I cannot see the commit in any branch.

**Comment by Isalia20** [comment]
@zap150 in pytorch, PRs do not get directly merged, they get merged by pytorchbot and then closed. PR has a label "Merged" so it is on main. To use it before it gets released, you can pull the nightly version of torch on your mac with:
`pip3 install --pre torch torchvision --index-url https://download.pytorch.org/whl/nightly/cpu`

**Comment by Isalia20** [comment]
You can see the diff in the PR and the code on main here:
https://github.com/pytorch/pytorch/blob/d7a8e8282c1afb2d52b07b13f54a3341848a5b3c/aten/src/ATen/native/mps/operations/Distributions.mm#L571-L584
